base/optional_serialization.hpp

```diff
@@ -0,0 +1,31 @@
+
+#pragma once
+
+#include <experimental/optional>
+#include <functional>
+#include <type_traits>
+
+namespace principia {
+namespace base {
+
+template<typename Message>
+struct OptionalField {
+  std::function<Message*()> mutable_field;
+  std::function<void()> clear_field;
+};
+
+#define OPTIONAL_FIELD(message, field)                      \
+  (::principia::base::OptionalField<                        \
+       std::remove_cv_t<std::remove_reference_t<            \
+           decltype(message->field())>>> {                  \
+       [m = (message)]() { return m->mutable_##field(); },  \
+       [m = (message)]() { m->clear_##field(); }})
```

Comment:
This business of going through lambdas seems unnecessarily convoluted.  You could just have a 3-argument macro:
`#define WRITE_TO_OPTIONAL(message, field, value)` and it would not be less readable or less safe.  Heck, it could even have only 2 arguments if you ensure that `field` and `value` have the same name.